services/webnn/ort/graph_builder_ort.cc

```diff
@@ -258,6 +259,658 @@ GraphBuilderOrt::GraphBuilderOrt(
+      ++operand_id_to_use_count_map[input_id];
+    }
+
+    // Try to find dequantizeLinear that can be fused into following matmul
+    // operations.
+    switch (operation->which()) {
+      case mojom::Operation::Tag::kMatmul: {
+        // Map matmul's inputs to operation, so the following algorithm can find
+        // a dequantizeLinear whose output is consumed by a matmul.
+        CHECK_EQ(operation_connectivity.input_ids.size(), 2U);
+        // We needn't check the result of `try_emplace` here, because if the key
+        // `input_id` is already in container, there must be more than 1 output
+        // edges from a predecessor in which case the dequantizeLinear fusion
+        // won't happen.
+        input_id_to_matmul_map.try_emplace(operation_connectivity.input_ids[1],
+                                           operation.get());
+        break;
+      }
+      case mojom::Operation::Tag::kTranspose: {
+        // If a transpose's output is solely used by a matmul and
+        // transpose's output must be the 2nd input of a matmul.
+        CHECK_EQ(operation_connectivity.output_ids.size(), 1U);
+        uint64_t output_id = operation_connectivity.output_ids[0];
+        // MatMul has only 1 transpose input and the tranpose must have 1 output
+        // edge and not be a graph output
+        if (!input_id_to_matmul_map.contains(output_id) ||
+            operand_id_to_use_count_map[output_id] != 1) {
+          break;
+        }
+        input_id_to_pre_matmul_transpose_map.try_emplace(
+            operation_connectivity.input_ids[0], operation.get());
+        break;
+      }
+      case mojom::Operation::Tag::kReshape: {
+        CHECK_EQ(operation_connectivity.output_ids.size(), 1U);
+        uint64_t output_id = operation_connectivity.output_ids[0];
+        if ((!input_id_to_pre_matmul_transpose_map.contains(output_id) &&
+             !input_id_to_pre_matmul_transpose_reshape_dq_map.contains(
+                 output_id)) ||
+            operand_id_to_use_count_map[output_id] != 1) {
+          break;
+        }
+        //           reshape
+        //              |
+        //  input_a    transpose
+        //     \       /
+        //        matmul
+        if (input_id_to_pre_matmul_transpose_map.contains(output_id)) {
+          input_id_to_pre_matmul_transpose_reshape_map.try_emplace(
+              operation_connectivity.input_ids[0], operation.get());
+          break;
+        }
+        // Matmulnbits fusion pattern:
+        //            scale
+        //              |
+        //     input reshape_2  zeroPoint
+        //        \     |      /
+        //         dequantizeLinear
+        //              |
+        //           reshape_1
+        //              |
+        //  input_a    transpose
+        //     \       /
+        //        matmul
+        if (input_id_to_pre_matmul_transpose_reshape_dq_map.contains(
+                output_id)) {
+          // Add reshape_2
```

Comment:
Append period at the end of the sentence. Do the same for the following comments.
```suggestion
          // Add reshape_2.
```